protocol_details/rpmsg_protocol.rst

```diff
@@ -0,0 +1,89 @@
+.. _rpmsg-protocol-work-label:
+
+=====================
+RPMsg Protocol Layers
+=====================
+
+The whole communication implementation can be separated in three different ISO/OSI layers - Transport, Media Access Control and Physical layer. Each of them can be implemented separately and for example multiple implementations of the Transport Layer can share the same implementation of the MAC Layer (VirtIO) and the Physical Layer. Each layer is described in following sections.
+
+.. image:: ../images/protocol_layers.jpg
+
+Physical Layer – Shared Memory
+------------------------------
+
+The solution proposed in this document requires only two basic hardware components - shared memory (accessible by both communicating sides) and inter-core interrupts (in a specific configuration optional). The minimum configuration requires one interrupt line per communicating core meaning two interrupts in total. This configuration is briefly presented in figure at the beginning of this section. It is to be noticed that no inter-core synchronization hardware element such as inter-core semaphore, inter-core queue or inter-core mutex is needed! This is thanks to the nature of the `virtqueue <https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.html#x1-270006>`_, which uses single-writer-single-reader circular buffering. (As defined in next subsection).
+
+
+The use of interrupts by the ring buffers is optional and can be suppressed using a bit set in their configuration flags field - in such a configuration, the interrupts are not necessary. However both cores need to poll these ring buffers for new incoming messages, which may not be optimal.
+
+.. image:: ../images/core_to_core_interrupt.jpg
+
+Media Access Layer - VirtIO
+---------------------------
+
+This layer is the key part of the whole solution - thanks to this layer, there is no need for inter-core synchronization. This is achieved by a technique called single-writer single-reader circular buffering, which is a data structure enabling multiple asynchronous contexts to interchange data.
+
+.. image:: ../images/vrings.jpg
+
+This technique is however applicable only in core-to-core configuration, not in core-to-multicore configuration, since in such a case, there would be multiple writers to the “IN” ring buffer. This would require a synchronization element, [such as a semaphore?], which is not desirable.
+
+The above shown picture describes the vring component. Vring is composed of three elementary parts - buffer descriptor pool, the “available” ring buffer (or input ring buffer) and the “used” ring buffer (or free ring buffer). All three elements are physically stored in the shared memory.
```

Comment:
"input ring buffer" and "free ring buffer" is not really current terminology in Virtio lingo. Maybe just keep the standard used and available terminology.